### ISSUE TITLE
Parser for eml files?

### ISSUE THREAD

**Issue by piernik** [ddeboer/imap#265]
Is it possible to use Your library as parser of eml files?
Now I'm using https://github.com/php-mime-mail-parser/php-mime-mail-parser

But it would be great to use only one library.

**Comment by Slamdunk** [comment]
Theoretically yes, an EML file can be treated like an EmbeddedMessage.

Would you be so kind to create PR containing a proof-of-concept?

**Comment by piernik** [comment]
I think It's not possible right now. `EmbeddedMessage` extends `AbstractPart` at the end which needs Imap resource as constructor param. All interesting methods are in `AbstractPart` class (as I think).

**Comment by Slamdunk** [comment]
You are right, I was referring that there is only one raw imap call in the whole EmbeddedMessage logic to serve the BasicMessageInterface API, and that call is easily overwritable to work with an EML file.

**Comment by piernik** [comment]
So ti's not possible to use those methods without IMAP resource? Are You planning to adjust Your lib to use eml parser?

**Comment by Slamdunk** [comment]
It is possible, and I think pretty straightforward, to get what you ask.

Not soon though. I'll tag it in 1.2 milestone.

**Comment by piernik** [comment]
Would be great.

**Comment by piernik** [comment]
I've uploaded parser: https://github.com/ddeboer/imap/pull/268

**Comment by piernik** [comment]
Whe do You expect deliver it??

**Comment by Slamdunk** [comment]
I don't know yet, I have personal issues that keep me off coding.

**Comment by piernik** [comment]
maybe a timeframe? I need to publish my project by the end of the year so by 15th I need to start testing. Will it be ready within 2 weeks??

**Comment by Slamdunk** [comment]
This is a free and open source software developed for free and maintained in our spare time, we can't guarantee any timeframe, so we don't give them.

**Comment by piernik** [comment]
I understad - I'll use different parser :/

**Comment by piernik** [comment]
Any news on that?

**Comment by pyatnitsev** [comment]
Is it actual? Do you need some help with PR?

**Comment by Slamdunk** [comment]
I'd like to be honest: I'm not working on this and I haven't even planned to.

Any PR is welcome: I think it should be easy to get a parser that returns a `BasicMessageInterface` implementation for a generic message.

Tests are mandatory: start the basic feature against `tests/fixtures/plain_only.eml` and submit a PR, after discussing the general approach a more comprehensive data set can be tested too.

**Comment by pyatnitsev** [comment]
Ok. Lets discuss an implementation of this feature. 
I see it, as a separate class, maybe EmlParser with method parseEml() that have one param - path to eml object amd returns instance of new class - ParsedMessage (implements BasicMessageInterface)

right?

**Comment by pyatnitsev** [comment]
Another idea: are you realy want to get own implementation of eml parser, maybe better will be integration of https://github.com/php-mime-mail-parser/php-mime-mail-parser to get BasicMessageInterface object? 

WDYT?

**Comment by Slamdunk** [comment]
https://github.com/php-mime-mail-parser/php-mime-mail-parser is a front-end for the [Mailparse](https://www.php.net/manual/en/book.mailparse.php) extension.

I think it would be good to use that extension too.

A less pretty but still valid idea is to provide an EML parser that just parses the headers with [imap_rfc822_parse_headers](https://www.php.net/manual/en/function.imap-rfc822-parse-headers.php) function, leaving the raw body. This wouldn't require any other third-party library/extension.

**Comment by rikvdh** [comment]
@Slamdunk I was working on this as I was migrating from laminas-mail to here and I need this for testing.

I am quite far with this, some 'final' keywords in the AbstractMessage need to go in order to override them in a 'EmlMessage' class. It can decode headers now, no problem. But when decoding content, it relies on a `structure`-object generated by `imap_fetchstructure`. 

I see a way with php-mime-mail-parser/php-mime-mail-parser or using mailparse directly. But is that desirable? I don't want to put effort in making it work when that route is not 'the way to go forward'.

See https://github.com/ddeboer/imap/compare/master...rikvdh:imap:feature/eml-import?expand=1
